src/server/services/validate-upload.js

```diff
@@ -1,105 +1,82 @@
-/* eslint camelcase: 0 */
 
-let log = () => {}
-if (process.env.VERBOSE) {
-  log = console.dir
-}
-
-const xlsx = require('xlsx')
-const {
-  currentReportingPeriodSettings,
-  getPriorPeriodSummaries
-} = require('../db')
-
-const reportingPeriods = require('../db/reporting-periods')
-
-const { getValidationTemplateSheets } = require('./get-template')
-const { parseFilename } = require('./parse-filename')
-const { ValidationLog } = require('../lib/validation-log')
-const { validateData } = require('./validate-data')
-const {
-  parseSpreadsheet,
-  removeSourceRowField,
-  spreadsheetToDocuments
-} = require('../lib/spreadsheet')
-
-const { removeEmptyDocuments } = require('../lib/remove-empty-documents')
-
-const validateUpload = async ({
-  filename,
-  user_id,
-  agency_id, // eslint-disable-line
-  data,
-  reporting_period_id
-}) => {
-  if (!reporting_period_id) {
-    const period = await currentReportingPeriodSettings()
-    reporting_period_id = period.id
+const { documentsForUpload } = require('../db/documents')
+const { setAgencyId } = require('../db/uploads')
+const { agencyByCode } = require('../db/agencies')
+
+class ValidationError extends Error {
+  constructor (message, { severity = 1, tab = null, row = null, col = null } = {}) {
+    super(message)
+    this.severity = severity
+    this.tab = tab
+    this.row = row
+    this.col = col
   }
-  const reportingPeriod = await reportingPeriods.get(reporting_period_id)
 
-  const valog = new ValidationLog()
-  const { valog: filenameValog, ...fileParts } =
-    await parseFilename(filename, reportingPeriod)
-
-  valog.append(filenameValog)
-  if (!valog.success()) {
-    log('failed to validate file name')
-    return { valog, documents: {} }
+  toObject () {
+    return {
+      message: this.message,
+      severity: this.severity,
+      tab: this.tab,
+      row: this.row,
+      col: this.col
+    }
   }
+}
 
-  let workbookXlsx
-  try {
-    workbookXlsx = xlsx.read(data, { type: 'buffer' })
-  } catch (e) {
-    console.log('error', e)
-    valog.append(`Can't parse xlsx file ${filename}`)
-    return { valog, documents: {} }
-  }
+async function validateAgencyId ({ upload, documents }) {
+  // grab agency id from the cover sheet
+  const coverSheet = documents.find(doc => doc.type === 'cover').content
+  const agencyCode = coverSheet[1][0]
```

Comment:
@ajhyndman example of how to pull data out of the spreadsheet now that it's just an array of arrays, the error below can point to the same location as where i'm looking

Reply:
If we do go this route, I think I'd rather save the data in some format with richer metadata than array of arrays.  Trying to access or manipulate data by column and row position sounds pretty error-prone to me.

I believe sheetjs (XLSX library) has support for a "JSON" data type as well.  This saves rows as objects with named keys (derived from headers).  The API is weird, from reading the docs, I think you _omit_ the header option from sheet_to_json to get this behaviour?

Alternatively, we could probably just save the default sheetjs interchange format?